### PR TITLE
Add topk op

### PR DESCRIPTION
**Pull request by typhoonzero**
Fix https://github.com/PaddlePaddle/Paddle/issues/3759

Still got problems when using a 3d tensor input.

### CODE REVIEW

File: python/paddle/v2/framework/tests/CMakeLists.txt [hunk 1]
```diff
@@ -16,6 +16,7 @@ py_test(test_cross_entropy_op SRCS test_cross_entropy_op.py)
 py_test(test_gather_op SRCS test_gather_op.py)
 py_test(test_scatter_op SRCS test_scatter_op.py)
 py_test(test_fill_zeros_like_op SRCS test_fill_zeros_like_op.py)
+py_test(test_top_k_op SRCS test_top_k_op.py)
```
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/CMakeLists.txt]: do not find this file

Comment (typhoonzero) [reply python/paddle/v2/framework/tests/CMakeLists.txt]: Sorry didn't get your point?

**Comment by typhoonzero** [comment]
I'm looking into http://www.math.grin.edu/%7Eblanchaj/Research/ABGS_KSelection.pdf for some details of the performance of topk kernel.

File: paddle/operators/top_k_op.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    auto* indices = ctx.Output<Tensor>("Indices");
+    // k is determined by Attr
+    const size_t k = static_cast<AttrType>(ctx.op_.GetAttr<AttrType>("k"));
+
+    output->mutable_data<T>(ctx.GetPlace());
+    indices->mutable_data<T>(ctx.GetPlace());
+
+    auto X = EigenMatrix<T>::From(*input);
+    auto Out = EigenMatrix<T>::From(*output);
+    auto Indices = EigenMatrix<T>::From(*indices);
```
Comment (qingqing01) [inline paddle/operators/top_k_op.h]: If using the `std::partial_sort` and not using the function in Eigen, there is no need to convert to Eigen tensor.

File: paddle/operators/top_k_op.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    auto place = ctx.GetEigenDevice<Place>();
+    Out.device(place);
+    Indices.device(place);
+
+    for (size_t i = 0; i < row; i++) {
+      // TODO(typhoonzero): make this more efficient
+      std::vector<std::pair<T, size_t>> vec;
+      for (size_t j = 0; j < col; j++) {
+        vec.push_back(std::pair<T, size_t>(X(i, j), j));
```
Comment (qingqing01) [inline paddle/operators/top_k_op.h]: Maybe we can use the plain pointer to get data and there is no need to use Eigen.

File: paddle/operators/top_k_op.cu [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+  __device__ __forceinline__ bool operator<(const Pair<T>& in) const {
+    return (v_ < in.v_) || ((v_ == in.v_) && (id_ > in.id_));
+  }
+
+  __device__ __forceinline__ bool operator>(const Pair<T>& in) const {
+    return (v_ > in.v_) || ((v_ == in.v_) && (id_ < in.id_));
+  }
+
+  T v_;
+  int id_;
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/top_k_op.cu]: > Data members of classes, both static and non-static, are named like ordinary nonmember variables, but with a trailing underscore.

Seems underscore is still needed.

File: paddle/operators/top_k_op.cu [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    if (k > input_width) k = input_width;
+
+    // pass lds and dim same to input width.
+    // NOTE: old matrix implementation of stride is different to eigen.
+    // TODO(typhoonzero): launch kernel on specified stream, need sync?
+    // TODO(typhoonzero): refine this kernel.
+    dim3 threads(256, 1);
+    dim3 grid(input_height, 1);
+
+    KeMatrixTopK<T, 5, 256><<<grid, threads>>>(
```
[the remark this answers is not in the document]
Comment (typhoonzero) [reply paddle/operators/top_k_op.cu]: `ctx.device_context()` returns `const` type currently, can not be `reinterpret_cast`ed currently?

File: python/paddle/v2/framework/tests/CMakeLists.txt (hunk 1 above)
Comment (typhoonzero) [reply python/paddle/v2/framework/tests/CMakeLists.txt]: Thanks. Added.

File: paddle/operators/top_k_op.h (hunk 3 above)
Comment (typhoonzero) [reply paddle/operators/top_k_op.h]: Follow comments, all done. Please review again, Thanks very much!

File: paddle/operators/top_k_op.cc [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+class TopkOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(const framework::InferShapeContext &ctx) const override {
+    PADDLE_ENFORCE_NOT_NULL(ctx.InputVar("X"),
+                            "Input of TopkOP must be initialized.");
+    auto *input = ctx.Input<framework::Tensor>("X");
+    const int k = static_cast<int>(ctx.op().GetAttr<int>("k"));
```
Comment (qingqing01) [inline paddle/operators/top_k_op.cc]: ctx.op().GetAttr ->ctx.GetAttr.

if merged after this PR: https://github.com/PaddlePaddle/Paddle/pull/3903
please use Attr.

File: paddle/operators/top_k_op.cc [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    // k must >= 1
+    PADDLE_ENFORCE_GE(k, 1);
+    // input must have >= 1d shape.
+    PADDLE_ENFORCE_GE(input->dims().size(), 1);
+    // input must have >= k columns.
+    PADDLE_ENFORCE_GE(input->dims()[input->dims().size() - 1], k);
```
Comment (qingqing01) [inline paddle/operators/top_k_op.cc]: need comments in  PADDLE_ENFORCE_GE

File: paddle/operators/top_k_op.cu [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+    // pass lds and dim same to input width.
+    // NOTE: old matrix implementation of stride is different to eigen.
+    // TODO(typhoonzero): launch kernel on specified stream, need sync?
+    // TODO(typhoonzero): refine this kernel.
+    dim3 threads(256, 1);
+    dim3 grid(input_height, 1);
+
+    // auto stream = reinterpret_cast<platform::CUDADeviceContext*>(
+    // ctx.device_context())->stream();
```
Comment (qingqing01) [inline paddle/operators/top_k_op.cu]: please refer this PR: https://github.com/PaddlePaddle/Paddle/pull/3753/files#diff-be198775c4ad57a522949cb9ff47c6b2R89

File: paddle/operators/top_k_op.h [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+    auto X = EigenMatrix<T>::From(*input);
+
+    // reshape input to a flattern matrix(like flat_inner_dims)
+    framework::DDim inputdims = input->dims();
+    const size_t row = framework::product(
+        framework::slice_ddim(inputdims, 0, inputdims.size() - 1));
+    const size_t col = inputdims[inputdims.size() - 1];
+    Eigen::DSizes<int, 2> flat2dims(row, col);
+    X.reshape(flat2dims);
```
Comment (qingqing01) [inline paddle/operators/top_k_op.h]: X -> x. all variable names should be lowercase.

File: paddle/operators/top_k_op.h [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    T* indices_data = indices->mutable_data<T>(ctx.GetPlace());
+
+    auto X = EigenMatrix<T>::From(*input);
+
+    // reshape input to a flattern matrix(like flat_inner_dims)
+    framework::DDim inputdims = input->dims();
+    const size_t row = framework::product(
+        framework::slice_ddim(inputdims, 0, inputdims.size() - 1));
+    const size_t col = inputdims[inputdims.size() - 1];
+    Eigen::DSizes<int, 2> flat2dims(row, col);
```
Comment (qingqing01) [inline paddle/operators/top_k_op.h]: 我觉得通常是： row = inputdims[0], col = product(inputdims[1: ...])

File: paddle/operators/top_k_op.h [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+template <typename T, int MajorType = Eigen::RowMajor,
+          typename IndexType = Eigen::DenseIndex>
+using EigenMatrix = framework::EigenMatrix<T, MajorType, IndexType>;
+
+template <typename Place, typename T>
+class TopkKernel : public framework::OpKernel {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    // Get the top k elements of each row of input tensor
+    // FIXME: only deal with matrix(2d tensor).
```
Comment (qingqing01) [inline paddle/operators/top_k_op.h]: 实际上我觉得当前可以只支持rank=2的，用户如果需要可以TopK之前加reshap op.

Comment (typhoonzero) [reply paddle/operators/top_k_op.h]: 嗯嗯，同感可以先merge后续增加。

File: paddle/operators/top_k_op.h (hunk 10 above)
Comment (typhoonzero) [reply paddle/operators/top_k_op.h]: 这应该是两种不同的flattern的方法，我提过一个issue：https://github.com/PaddlePaddle/Paddle/issues/3771， 参考TF的实现是这样的。

**Pull request merged**